scvi/external/cell2location/_model.py

```diff
@@ -0,0 +1,572 @@
+        :param iter_end: omit last iterations from the plot
+        """
+
+        if ax is None:
+            ax = plt
+            ax.set_xlabel = plt.xlabel
+            ax.set_ylabel = plt.ylabel
+
+        if history_key is None:
+            history_key = self.history.keys()
+
+        if type(history_key) == str:
+            history_key = [history_key]
+
+        for i in history_key:
+
+            if iter_end == -1:
+                iter_end = np.array(self.history[i]).flatten().shape[0]
+
+            y = np.array(self.history[i]).flatten()[iter_start:iter_end]
+            if log_y:
+                y = np.log10(y)
+            ax.plot(np.arange(iter_start, iter_end), y, label='train')
+            ax.set_xlabel('Training epochs')
+            ax.set_ylabel('Reconstruction accuracy (ELBO loss)')
+            ax.legend()
+            plt.tight_layout()
+            
+            
+    def export2adata(self, adata, slot_name='mod'):
+        r""" Add posterior mean and sd for all parameters to unstructured data `adata.uns['mod']`.
+
+        :param adata: anndata object
+        """
+        # add factor filter and samples of all parameters to unstructured data
+        adata.uns[slot_name] = {}
+
+        adata.uns[slot_name]['mod_name'] = str(self.__class__.__name__)
+        adata.uns[slot_name]['fact_filt'] = self.fact_filt
+        adata.uns[slot_name]['fact_names'] = self.fact_names.tolist()
+        adata.uns[slot_name]['var_names'] = self.var_names.tolist()
+        adata.uns[slot_name]['obs_names'] = self.obs_names.tolist()
+        adata.uns[slot_name]['post_sample_means'] = self.samples['post_sample_means']
+        adata.uns[slot_name]['post_sample_sds'] = self.samples['post_sample_sds']
+        adata.uns[slot_name]['post_sample_q05'] = self.samples['post_sample_q05']
+        adata.uns[slot_name]['post_sample_q95'] = self.samples['post_sample_q95']
+
+        return adata
+
+
+class Cell2locationModelClass(Cell2locationBaseModelClass):
+    """
+    Reimplementation of cell2location [Kleshchevnikov20]_ model. ModelClass. 
+
+    https://github.com/BayraktarLab/cell2location
+
+    Parameters
+    ----------
+    sc_adata
+        single-cell AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+    use_gpu
+        Use the GPU or not.
+    **model_kwargs
+        Keyword args for :class:`~scvi.external.cell2location...`
+
+    Examples
+    --------
+    >>> 
+    """
+
+    def __init__(
+        self,
+        adata: AnnData,
+        cell_state_df: pd.DataFrame,
+        var_names_read = None,
+        sample_id = None,
+        use_gpu: bool = True,
+        batch_size: int = 1024,
+        **model_kwargs,
+    ):
+        
+        intersect = np.intersect1d(cell_state_df.index, adata.var_names)
+        cell_state_df = cell_state_df.loc[intersect,:]
+        adata = adata[:, intersect]
+        adata.varm['cell_state_varm'] = cell_state_df.values
```

Comment:
what is this?

Reply:
This subsets adata and `cell_state_df` with signatures of cell types to common genes. Doing that in the notebook does not guarantee that users will always do it or do it correctly.
I can add a error instead

```python
all_matching = np.all(cell_state_df.index == adata.var_names)
if not all_matching:
    raise ValueError('cell_state_df and spatial adata should have the same genes')
```